### ISSUE TITLE
Error when run optimizer.compute_gradients

### ISSUE THREAD

**Issue by gliese581gg** [tensorflow/tensorflow#1470]
I'm trying to do training with multiple threads

I tried to run optimizer.minimize() in each threads but it didn't work because threads overwrited each other's weight updates. So I'm trying to stack computed gradients in each threads and apply mean or sum of them in another thread.
So I made a code like

w1 = tf.Variable(tf.random_normal([shape], stddev=0.01))
b1 = tf.Variable(tf.constant(0.1, shape=[filters]))
rmsprop = tf.train.RMSPropOptimizer(lr,decay,0.0,eps)
grads = rmsprop.compute_gradients(cost,[w1,b1])
apply_grads = rmsprop.apply_gradients(grads)

If i run apply_grads, it works fine. but when i run grads to get gradients, error occurs like below.
`[(<tf.Tensor ~~ shape=~~ dtype=float32>, <tensorflow.python.ops.variables.Variable object at 0x7f6da5fd7350>)] of [(<tf.Tensor ~~ shape=~~ dtype=float32>, <tensorflow.python.ops.variables.Variable object at 0x7f6da5fd7350>)]  has invalid type <type "list">, must be a string or Tensor. (Can not convert a list into a Tensor or Operation.)`

How can I get gradients for optimizer?

Or is there a efficient way to do training with multiple threads?

**Comment by mrry** [comment]
Are you passing `grads` directly to `Session.run()`? I believe that will contain a list of (gradient, variable) tuples, which needs to converted to a list a tensors in order to pass it to `Session.run()`. For example, you could do:

```python
grads_and_vars = rmsprop.compute_gradients(cost, [w1, b1])
grads = [g for g, _ in grads_and_vars]
vars = [v for _, v in grads_and_vars]
grad_vals = sess.run(grads)
var_to_grad_val_map = {v.name: val for (v, val) in zip(vars, grad_vals)}
```

**Comment by rinnadom** [comment]
Was this the correct answer? I am having the same problem

**Comment by jediweon** [comment]
I'm facing the same problem. Fetching the gradient values with session.run() always result in Type Error - which was mentioned by @gliese581gg .
I've tried various terms, including code snippet @mrry suggested above, but not so effective.

**Comment by jediweon** [comment]
I'm facing the same problem. Fetching the gradient values with session.run() always result in Type Error - which was mentioned by @gliese581gg .
I've tried various terms, including code snippet @mrry suggested above, but not so effective.

**Comment by jediweon** [comment]
Now I see the problem. With your program ended without freeing related resources, Variables with trainable=True will remain and be called by optimizer's compute_gradients() unless you specify Variables you want to know about. And that will result in returning non value objects(which is NonType), for those remaining Variables has nothing to do with current program.

This can be avoided in various ways - like calling tf.reset_default_graph() - , but still, I believe this is non-intentional bug, and should be fixed soon enough.

**Comment by wangxiaodong1021** [comment]
When I get the gradient of the InceptionV4 model, I get the same problem. The value of None appears in each BatchNorm layer, I'm not sure if this is an bug. Or, the BatchNorm layer would have no gradient.
